src/lib/src/regsurf/utilities.cpp

```diff
@@ -230,4 +230,72 @@ find_cell_range(const double xmin,
     return { i_min, i_max, j_min, j_max };
 }
 
+/*
+ * Function to get the Z value at a given X, Y position on a regular 2D grid
+ *
+ * @param x The x-coordinate of the point
+ * @param y The y-coordinate of the point
+ * @param xori The x-coordinate of the regsurf origin
+ * @param yori The y-coordinate of the regsurf origin
+ * @param xinc The x-increment
+ * @param yinc The y-increment
+ * @param ncol The number of columns
+ * @param nrow The number of rows
+ * @param angle_deg The angle of rotation in degrees, anticlockwise from X
+ * @param values The 2D array of values on the regsurf (where Nan values are masked)
+ * @return The interpolated Z value at the given X, Y position
+ */
+
+double
+get_z_from_xy(const double x,
+              const double y,
+              const double xori,
+              const double yori,
+              const double xinc,
+              const double yinc,
+              const size_t ncol,
+              const size_t nrow,
+              const double angle_deg,
+              const py::array_t<double> &values)
+{
+    // Convert the angle to radians
+    double angle_rad = angle_deg * M_PI / 180.0;
+
+    Point p = { x, y, 0 };
+    Point p_rel = inverse_rotate_and_translate(p, xori, yori, std::cos(angle_rad),
+                                               std::sin(angle_rad));
+
+    // Find the indices of the grid cell containing the point
+    int i = static_cast<int>(p_rel.x / xinc);
+    int j = static_cast<int>(p_rel.y / yinc);
+
+    // Check if the point is outside the grid, and return NaN if it is
+    if (i < 0 || i >= static_cast<int>(ncol - 1) || j < 0 ||
+        j >= static_cast<int>(nrow - 1)) {
+        return std::numeric_limits<double>::quiet_NaN();
+    }
```

Comment:
```suggestion
    size_t i = static_cast<size_t>(p_rel.x / xinc);
    size_t j = static_cast<size_t>(p_rel.y / yinc);

    // Check if the point is outside the grid, and return NaN if it is
    if (i < 0 || i >= static_cast<size_t>(ncol - 1) || j < 0 ||
        j >= static_cast<size_t>(nrow - 1)) {
        return std::numeric_limits<double>::quiet_NaN();
    }
```
Always safer to use an unsigned integer for buffer indices

Reply:
I think that will change the math, since the check for a point being outside is _that i or j may be negative_. I propose to apply an temp. `int` and the cast to `size_t` after the test:

```
    // Find the indices of the grid cell containing the point
    int i_temp = static_cast<int>(p_rel.x / xinc);
    int j_temp = static_cast<int>(p_rel.y / yinc);

    // Check if the point is outside the grid, and return NaN if it is
    if (i_temp < 0 || i_temp >= static_cast<int>(ncol - 1) || j_temp < 0 ||
        j_temp >= static_cast<int>(nrow - 1)) {
        return std::numeric_limits<double>::quiet_NaN();
    }

    // Convert to size_t after validation
    size_t i = static_cast<size_t>(i_temp);
    size_t j = static_cast<size_t>(j_temp);
    // Get the values at the corners of the cell
```

Reply:
Oops, you're right, I missed that part... which would've been worse